Port to GIO and debhelper 10.

parazyd opened the pull request:
HAM now builds in Devuan Beowulf/Debian Buster.
I tested saving the logfile, and installing a package and both seemed to work fine.

Obviously more work is needed, but it's a good base, and lets us move to Beowulf.

parazyd commented:
Note that I did not rename some function calls like `annoy_user_with_gnome_vfs_result`.

Some things that I couldn't completely figure out and would like help with:

* https://github.com/maemo-leste/hildon-application-manager/pull/1/files#diff-005d118f141929270b3e1c4150887ae2R866
* https://github.com/maemo-leste/hildon-application-manager/pull/1/files#diff-005d118f141929270b3e1c4150887ae2R3392

There are also some TODOs I left as comments, and some old code that I commented and left there for reference.

Appreciate all input :)

spinal84 commented:
Doing such quickie hacky commits forces someone to do double work in the future.
I may be wrong, but it's better to do pull requests for more or less complete work.

Making a package to just build and then commit the quick dirty fixes is not a good approach, because if someone (not you) will try to maintain the package in the future, your work will just be abandoned and redone from scratch. Because such things are very hard to support. It means your time spent on these commits will be wasted and the time of another developer who will try to maintain the package will be wasted as well.

Tip one: understanding how things work before removing something is a must.
Tip two: detailed explanations of the reasoning why some changes are needed are always good for commits history.
Tip three: if there are a lot of non-trivial changes in a single commit, it's often a good idea to split that commit to several simpler commits.

Please get it correct: having a good commits history simplifies the developers' life in the future.

freemangordon commented on src/util.cc:
```diff
@@ -3493,19 +3492,16 @@ localize_file_and_keep_it_open (const char *uri,
+  GFileInputStream *file;
+  GError *error = NULL;
 
   if (vfs_uri == NULL)
```
this can never be NULL, so you can remove that check (and error handling) altogether

freemangordon commented on src/util.cc:
```diff
@@ -3515,23 +3511,23 @@ localize_file_and_keep_it_open (const char *uri,
-	call_copy_cont (result);
+      if (file == NULL)
+      {
+	    call_copy_cont (error);
```
identation?

freemangordon commented on src/util.cc:
```diff
@@ -3515,23 +3511,23 @@ localize_file_and_keep_it_open (const char *uri,
-	  call_copy_cont (GNOME_VFS_OK);
-	}
+	  {
+        copy_target = g_file_get_path (vfs_uri);
```
ditto

freemangordon commented on src/util.cc:
```diff
@@ -3551,25 +3547,29 @@ localize_file_and_keep_it_open (const char *uri,
 	{
 	  add_log ("Can not create %s: %m", copy_tempdir);
-	  call_copy_cont (GNOME_VFS_ERROR_GENERIC);
+      error = g_error_new (G_IO_ERROR, G_IO_ERROR_FAILED, "Can not create %s: %m", copy_tempdir);
```
I think you got it :)

freemangordon commented on src/util.cc:
```diff
@@ -3580,11 +3580,13 @@ cleanup_temp_file ()
   */
 
+  /*
   if (copy_vfs_handle)
```
why keeping that code commented?

freemangordon commented on src/instr.cc:
```diff
@@ -526,15 +526,13 @@ execute_card_install (GKeyFile *keyfile, const char *entry,
     }
 
   gchar *dirname = g_path_get_dirname (filename);
-  gchar *esc_dirname = gnome_vfs_escape_string (dirname);
```
https://developer.gnome.org/glib/stable/glib-URI-Functions.html#g-uri-escape-string ?

freemangordon commented on src/log.cc:
```diff
@@ -56,45 +55,48 @@ save_log_cont (bool res, void *data)
 
+  GFile *file;
+  GFileOutputStream *stream;
+  gsize bytes;
```
I think it is better to make this local to the code block it is used. Same for the other vars (if any similar)

freemangordon commented on src/log.cc:
```diff
@@ -56,45 +55,48 @@ save_log_cont (bool res, void *data)
+                }
+            }
+        }
+      g_object_unref (stream);
```
missing new line

freemangordon commented on src/log.cc:
```diff
@@ -126,8 +128,9 @@ save_log_cont (bool res, void *data)
-  GnomeVFSResult result;
+  GFileInfo *info;
+  GFile *file;
+  GError *error;
```
GError *error = NULL; ?

MerlijnWajer replied:
Just want to highlight the importance of this, without this, your code will randomly segfault

freemangordon commented on src/util.cc:
```diff
@@ -833,34 +833,39 @@ annoy_user_with_errno (int err, const gchar *detail,
+  if (error)
+    add_log ("%s: %s\n", detail, error->message);
+  else {
+    // TODO: Look into everything that calls this function
```
Is it ever called without error being set?

freemangordon commented on src/util.cc:
```diff
@@ -833,34 +833,39 @@ annoy_user_with_errno (int err, const gchar *detail,
-    annoy_user (_FM ("sfil_ib_readonly_location"),
-                cont, data);
+  // else if (result == GNOME_VFS_ERROR_READ_ONLY_FILE_SYSTEM)
+  //   annoy_user (_FM ("sfil_ib_readonly_location"),
```
G_IO_ERROR_READ_ONLY ? not sure though but looks similar

freemangordon commented on src/util.cc:
```diff
@@ -3320,13 +3324,12 @@ pixbuf_from_base64 (const char *base64)
 */
 
-GnomeVFSResult copy_result;
+GError copy_result;
```
isn't that static? also, why not GError *

freemangordon commented on src/util.cc:
```diff
@@ -3351,17 +3354,18 @@ call_copy_cont (GnomeVFSResult result)
 
-      if (result == GNOME_VFS_ERROR_IO)
+      //if (result == GNOME_VFS_ERROR_IO)
+	  if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_CONNECTED)) // TODO: Maybe it's this one, maybe not
```
I think there are a couple of those (G_IO_ERROR_HOST_UNREACHABLE and the likes), you'd better handle them all.

freemangordon commented on src/util.cc:
```diff
@@ -3371,59 +3375,59 @@ call_copy_cont (GnomeVFSResult result)
+static void 
+copy_progress (goffset cur, goffset tot, gpointer cancellable)
+{
+  GError *error;
```
= NULL

freemangordon commented on src/util.cc:
```diff
@@ -3371,59 +3375,59 @@ call_copy_cont (GnomeVFSResult result)
     {
-      return !copy_cancel_requested;
+      /* TODO: Is this still necessary?
+       *
```
I don;t think any special obex handling shall take place here. So, not necessary.

freemangordon commented on src/util.cc:
```diff
@@ -3433,24 +3437,21 @@ cancel_copy (void *unused)
 
-  target_uri = gnome_vfs_uri_new (target);
+  target_uri = g_file_new_for_uri (target);
   if (target_uri == NULL)
```
can;t be NULL

freemangordon reviewed (changes requested):
Please, try to not mix C and C++ style comments, my eyes bleed everytime they see such a mixture :p

freemangordon commented on src/log.cc:
```diff
@@ -283,7 +283,7 @@ show_log_dialog_flow ()
 
-#if !HAVE_G_STRING_APPEND_VPRINTF
+#if 0 //!HAVE_G_STRING_APPEND_VPRINTF
 static void
```
could you explain (best in the commit message :) ) why is that needed?

parazyd commented:
@spinal84 Thanks for the input. I didn't plan to merge this PR as-is, rather to get your input and help.

I'll work more on it and split what makes sense into separate commits.

MerlijnWajer commented on configure.ac:
```diff
@@ -34,7 +34,7 @@ AC_EGREP_HEADER(apt_set_index_trust_level_for_package_hook,
 
 PKG_CHECK_MODULES(HAM_DEPS, glib-2.0 gtk+-2.0 hildon-1 hildon-fm-2 libosso
-                            conic gconf-2.0 gnome-vfs-2.0 mce libhildondesktop-1 x11)
+                            conic gconf-2.0 mce libhildondesktop-1 x11)
```
You'll want to add `gio-2.0` here

parazyd committed:
Port to GIO.

parazyd committed:
Port to debhelper 10.

parazyd commented:
Pushed tweaks. I'm not yet sure what to do with this: https://github.com/maemo-leste/hildon-application-manager/commit/efb446891cf9dca781d5bcdf94652ac0aebfc28e#diff-005d118f141929270b3e1c4150887ae2R3410

Regardless, things work.

parazyd committed:
Fix compile error.

log.cc:289:1: error: ‘void g_string_append_vprintf(GString*, const gchar*, __va_list_tag*)’ was declared ‘extern’ and later ‘static’ [-fpermissive]
 g_string_append_vprintf (GString *str, const gchar *fmt, va_list args)

spinal84 commented on debian/postinst:
```diff
@@ -68,6 +68,10 @@ hildon-application-manager-config update || true
 
+## HACK
+mkdir -p /usr/libexec
+ln -s /usr/lib/$(dpkg-architecture -q DEB_HOST_MULTIARCH)/apt-worker /usr/libexec/apt-worker
```
This command will fail when updating

spinal84 commented on debian/postinst:
```diff
@@ -68,6 +68,10 @@ hildon-application-manager-config update || true
 
+## HACK
+mkdir -p /usr/libexec
+ln -s /usr/lib/$(dpkg-architecture -q DEB_HOST_MULTIARCH)/apt-worker /usr/libexec/apt-worker
```
I think it should be done using configure.ac with AC_SUBST and friends.
It shouldn't be part of debian/ logic.

parazyd replied:
Could you help me with this please? I'm not well versed with autoconf.

freemangordon replied:
I think this should be done by using https://wiki.debian.org/Multiarch/Implementation#Dynamic_debian.2F.2A_files

parazyd replied:
@freemangordon The thing is that the path is hardcoded throughout h-a-m.

```
$ grep -RnisI libexec
ChangeLog:5678: * src/Makefile.am: Removed dist_libexec_SCRIPTS, since the
ChangeLog:5681: * src/main.cc: Use /usr/libexec/apt-worker.bin as the apt-worker
Makefile.am:18:rescuedir = $(libexecdir)
debian/hildon-application-manager.launcher:2:/usr/libexec/ham-after-boot
debian/hildon-application-manager.xsession:2:/usr/libexec/ham-after-boot &
debian/postinst:40:rem_inittab "resc::bootwait:/usr/libexec/apt-worker.bin rescue"
debian/postinst:44:rem_inittab "resc::bootwait:/usr/libexec/apt-worker rescue"
debian/postinst:72:mkdir -p /usr/libexec
debian/postinst:73:ln -s /usr/lib/$(dpkg-architecture -q DEB_HOST_MULTIARCH)/apt-worker /usr/libexec/apt-worker
debian/postrm:3:rm -f /usr/libexec/apt-worker
ham-rescue.sh:53:/usr/libexec/apt-worker rescue
hildon-application-manager.sudoers:1:%users ALL = NOPASSWD: /usr/libexec/apt-worker
src/Makefile.am:7:libexec_PROGRAMS = apt-worker ham-after-boot
src/apt-worker-client.cc:46:#define APT_WORKER_CMD_DEFAULT "/usr/libexec/apt-worker"
statusbar/ham-updates.c:457:    "/usr/libexec/apt-worker",
```

parazyd replied:
Guess I'll just learn how to substitute it with autoconf :)

spinal84 replied:
If you will have troubles with autoconf (I think you will), the mandatory thing here is to move this part out from debian/ directory. It should be done somewhere in Makefiles. So the package don't depend on Debian build system to be installed correctly.

parazyd replied:
Perhaps `ham-after-boot` and `apt-worker` should just be installed in `$(PREFIX)/bin`? If you take a look at any current debian-based system, there is never anything in `/usr/libexec`

parazyd replied:
Mitigated with e42f64940d62f34f87f7e0b334376bfb59a0f447 and 99bede06a39c29e13495113445f1d95ad38bf87e

spinal84 replied:
> Perhaps `ham-after-boot` and `apt-worker` should just be installed in `$(PREFIX)/bin`? If you take a look at any current debian-based system, there is never anything in `/usr/libexec`

I have a better option for you: take a look at how it's done in upower.
It hides its binary that shouldn't be executed by the user in /usr/lib/upower/.
As for me, it's a good example of how this should be done.
Maybe we can do the same for h-a-m?

parazyd replied:
I understand, but I am not sure how to replace the hardcoded strings in the c files.

EDIT: The only thing that comes to mind is patch the source through debian/patches, but that's ugly. With https://github.com/maemo-leste/hildon-application-manager/commit/e42f64940d62f34f87f7e0b334376bfb59a0f447 it ends up back in /usr/libexec.

spinal84 replied:
Just replace the hardcoded values in the sources by hand.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
-# Sample debian/rules that uses debhelper.
-# GNU copyright 1997 to 1999 by Joey Hess.
 
-# Uncomment this to turn on verbose mode.
```
I'd leave this comment untouched, because it comments the following line which you decided to leave.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
 
-CFLAGS = -Wall -Wno-format -g
+export DEB_CFLAGS_MAINT_APPEND  = -Wall -Wno-format -Wno-format-security -g -DTAP_AND_HOLD
+export DEB_LDFLAGS_MAINT_APPEND = -Wl,--as-needed
```
It's better to use "VAR += ..." expression and separated "export VAR" at the end of file.
This is useful because if someone adds new "VAR += ..." expression in the future, it will allow us to have good commit history for "git blame".

parazyd replied:
Ok, sure. Both have been used through our packaging, so I just put whatever came to mind.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
-	CFLAGS += -O0
-else
-	CFLAGS += -O2
-endif
```
These CFLAGS silently disappear in your commit.
It means the behavour of dpkg-buildpackage has changed.

parazyd replied:
It has, yes. It had changed in many packages ported to dh10 before.
We don't use this in Jenkins, and nobody had requested we keep it in the rules file so far.

spinal84 replied:
It's a separate action which should be committed separately.
It's not connected to "porting to debhelper 10".

parazyd replied:
Fine. I reverted it in 780bdf22fd425ba5acfd175baf70609250d7907c
For what I care it can just stay and we keep backwards compatibility.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
-	if [ ! -x configure ]; then ./autogen.sh; fi
-	dh_testdir
-	# Add here commands to configure the package.
-	CFLAGS="$(CFLAGS)" CXXFLAGS="$(CXXFLAGS)" ./configure --prefix=/usr --sysconfdir=/etc --mandir=\$${prefix}/share/man --infodir=\$${prefix}/share/info --disable-static $(CONFIGURE_OPTIONS)
```
Does this "--disable-static" change something?

parazyd replied:
The compilation creates a few static libraries along with the shared objects. Creating both is standard in Debian packages, and unless there is a strong reason not to build the static libraries, I would leave it like this.

freemangordon replied:
I'd rather keep --disable-static if .a is just build but not installed in the -dev package (which I suspect it isn't). Building static libs just increases compile time of the package.

parazyd replied:
It's installed in the -dev package. Reverted in 780bdf22fd425ba5acfd175baf70609250d7907c though. If the need arises, we can build static libraries again with a simple edit to debian/rules.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
-#	dh_installcron
-#	dh_installinfo
-#	dh_installman
-	dh_installxsession -u 'post 40'
```
Why is this line abandoned?

parazyd replied:
Sorry, yes, silently. The ham-after-boot program is used after system updates. It's not something that we do like Fremantle did, and should figure out if and how we want to use ham-after-boot.

I didn't think it was necessary, as it would be obvious that the xsession file is not installed.

spinal84 replied:
It would be good, if such things won't be messed with "Port to debhelper 10" commit.
It's a separate action which should be done in a separate commit wih enough explanation.

freemangordon replied:
I still don;t get it why we don;t do it like Fremantle did, care to elaborate?

parazyd replied:
Reverted in 5ae8c60060479faf6cd4c8569988960956d279bb 
You're right :)

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
-	dh_installxsession -u 'post 40'
-	dh_link
-	if test "x$(USE_MAEMO_LAUNCHER)" = "xtrue"; then \
-	  dh_maemolauncher -v; \
```
Why dh_maemolauncher is not used anymore?

parazyd replied:
AFAIK, it's used for injecting dependencies. We don't need this as they're properly handled in debian/control.

spinal84 replied:
What dependencies? Could you explain a bit more, what exactly it does?

parazyd replied:
https://github.com/maemo-leste/maemo-launcher/blob/master/debian/dh_maemolauncher

spinal84 replied:
Hm... What is this code for?
```
doit("mv", "-f", "$tmp/$so", "$tmp/$so.launch");
doit("ln", "-s", "/usr/bin/maemo-invoker", "$tmp/$so");
```

parazyd replied:
Reverted in 780bdf22fd425ba5acfd175baf70609250d7907c though. We can use it as then we'll have the .launch binary.

parazyd replied:
That code creates a .launch binary and instead launches binaries with maemo-invoker. See http://maemo.org/maemo_release_documentation/maemo4.1.x/node6.html#SECTION006311400000000000000

The same thing is happening in hildon-desktop, and IIRC @MerlijnWajer had some problems with starting h-d without maemo-invoker.

spinal84 replied:
Seems like we need it.

parazyd replied:
ACK. I reintroduced it and it'll be like it never went away when squashed :)

spinal84 commented on src/util.cc:
```diff
@@ -3549,27 +3526,32 @@ localize_file_and_keep_it_open (const char *uri,
+        add_log ("Can not chmod %s: %m", copy_tempdir);
+        error = g_error_new (G_IO_ERROR, G_IO_ERROR_FAILED, "Can not chmod %s: %m", copy_tempdir);
+        call_copy_cont (error);
+        g_error_free (error);
```
Poor logic.
If error return result is not used it should be replaced by NULL here.
call_copy_cont() should be rewritten accordingly.

spinal84 commented on src/main.cc:
```diff
@@ -2483,7 +2483,11 @@ install_from_file_flow (const char *filename,
 	  if (g_path_is_absolute (filename))
-	    fileuri = gnome_vfs_get_uri_from_local_path (filename);
+      {
+        GFile *file = g_file_new_for_path (filename);
+        fileuri = g_file_get_uri (file);
+        g_object_unref (file);
```
What's wrong with a single call to g_filename_to_uri() ?

parazyd replied:
I haven't tried it. It should work, just handle the error properly.

spinal84 commented on debian/control:
```diff
@@ -2,8 +2,8 @@ Source: hildon-application-manager
 Maintainer: Marius Vollmer <marius.vollmer@nokia.com>
-Build-Depends: debhelper (>= 4.0.0), libapt-pkg-dev (>= 0.7.6), libglib2.0-dev, libgtk2.0-dev, libhildon1-dev, libhildonfm2-dev, libconic0-dev, libgconf2-dev, libgnomevfs2-dev, mce-dev, libhildondesktop1-dev, libalarm-dev, libtime-dev, osso-af-settings, libcurl4-openssl-dev, maemo-launcher-dev, maemo-system-services-dev
-Standards-Version: 3.6.0
+Build-Depends: debhelper (>= 9), libapt-pkg-dev (>= 0.7.6), libglib2.0-dev, libgtk2.0-dev, libhildon1-dev, libhildonfm2-dev, libconic0-dev, libgconf2-dev, mce-dev, libhildondesktop1-dev, libalarm-dev, libtime-dev, osso-af-settings, libcurl4-openssl-dev, maemo-launcher-dev, maemo-system-services-dev
```
I may be wrong, but if you have debian/compat set to 10, you should have debhelper (>= 10) as well.

parazyd replied:
9 works with compat 10 as well.

spinal84 commented on src/log.cc:
```diff
@@ -57,46 +56,50 @@ save_log_cont (bool res, void *data)
-	      success = false;
-	    }
-	}
+      annoy_user_with_gnome_vfs_result (error, uri, save_log_do_nothing, NULL);
```
If we get rid of gnomevfs, this should be renamed.

parazyd commented:
Pushed more commits that should be autosquashed.

freemangordon commented on debian/rules:
```diff
@@ -1,114 +1,25 @@
 	USE_MAEMO_LAUNCHER = true
 endif
 
-CFLAGS += -DTAP_AND_HOLD
```
Are you sure this shall be removed?

spinal84 replied:
He moved it to CFLAGS_MAINT.
But I agree, it's better to leave it around for better observing what has changed.

parazyd replied:
Moved to CPPFLAGS_MAINT in 780bdf22fd425ba5acfd175baf70609250d7907c

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,29 @@
-endif
 ifeq (,$(findstring nolauncher,$(DEB_BUILD_OPTIONS)))
 	CONFIGURE_OPTIONS += --enable-maemo-launcher
 	USE_MAEMO_LAUNCHER = true
```
Where is USE_MAEMO_LAUNCHER used?

parazyd replied:
It is used now to check if we should run dh_maemolauncher.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,29 @@
+CFLAGS_MAINT  += -Wall -Wno-format -Wno-format-security -g -DTAP_AND_HOLD
+LDFLAGS_MAINT += -Wl,--as-needed
+
+export DEB_CFLAGS_MAINT_APPEND  = $(CFLAGS_MAINT)
```
AFAIK h-a-m is using c++ mostly

parazyd replied:
Saying it should be CXXFLAGS?

spinal84 replied:
I don't know. I only noticed that the flags you set were not used while the package was being built.

spinal84 replied:
I would check how the  build system before your porting was doing the things.
The FLAGS might be broken before you started on this. Who knows.

parazyd replied:
Added CXXFLAGS in 780bdf22fd425ba5acfd175baf70609250d7907c

parazyd commented:
Alright. Squashed and pushed. This should be it.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,50 @@
+	dh_install
+	dh_installxsession -u 'post 40'
+	if test "x$(USE_MAEMO_LAUNCHER)" = "xtrue"; then \
+		dh_maemolauncher -v; \
```
Indentation of this line differs from the following "for". This seems inconsistent.

parazyd replied:
Fixed.

spinal84 commented on src/log.cc:
```diff
@@ -139,33 +143,34 @@ save_log (char *uri, void *data)
         g_free (last_save_log_dir);
       last_save_log_dir = NULL;
 
       /* Annoy user with error */
```
This comment seems excessive now.

parazyd replied:
What do you mean?

spinal84 replied:
```
      /* Annoy user with error */
      annoy_user_with_gerror (error, uri, save_log_do_nothing, NULL);
```

This comment doesn't seem like something precious, when you're looking at the line it's dedicated for. Don't you think?

parazyd replied:
Just FYI, I really think nitpicking about comments that I didn't even touch is way out of scope for this pull request.

spinal84 replied:
If comments become useless/senseless due to changes in the code, this also needs to be fixed, imho.

parazyd replied:
No functionality has changed here.

parazyd replied:
Also let's please stop wasting time on this. I have better things to do.

spinal84 commented on src/log.cc:
```diff
@@ -139,33 +143,34 @@ save_log (char *uri, void *data)
       /* Annoy user with error */
-      annoy_user_with_gnome_vfs_result (result, uri,
-					save_log_do_nothing, NULL);
+      annoy_user_with_gerror (error, uri, save_log_do_nothing, NULL);
```
Maybe change it to annoy_user_with_error?

parazyd replied:
I wouldn't, since there are multiple `annoy_user` functions, and I'd rather keep it explicit.

spinal84 replied:
Ok, then, how about renaming it to annoy_user_with_g_error, to be consitent with how glib/gtk+ does with its functions/variables/types names?

There're some rules in translating mixed upper&lower case names to only lowercase.
In simple, it looks like this:
HelloWorld -> hello_world, GObject -> g_object, ViewerFile -> viewer_file, GType -> g_type, etc.

Take a look here for some live examples: https://developer.gnome.org/gobject/stable/gtype-conventions.html

In our case it would be:
GError -> g_error

parazyd replied:
Back in history, when h-a-m used to report errors with GError, the function was called `annoy_user_with_gerror` and I've kept it the same.

spinal84 replied:
As far as I can see annoy_user_with_gerror() was already declared in file src/instr.cc (as static) before your commits. What you did is overlapping this function with replaced one. BTW, src/instr.cc includes the file <util.h> where you declared "new" annoy_user_with_gerror() function. I may be wrong, but all this just seems a bit wrong to me.

This is from src/instr.cc:
```
static void
annoy_user_with_gerror (const char *filename, GError *error,
                        void (*cont) (void *), void *data)
```
This is from src/util.h (please, fix the indentation inconsistency you introduced):
```
void annoy_user_with_gerror (GError *error,
                                       const gchar *detail,
                                       void (*cont) (void *), void *data);
```

P.S. What "back in history" commit did you mean? I can see that annoy_user_with_gnome_vfs_result() was here since the first commit.

If you have some older sources, could you share a link please?

spinal84 commented on src/util.cc:
```diff
@@ -3459,20 +3452,18 @@ do_copy (const char *source, GnomeVFSURI *source_uri,
+                     copy_progress, // GFileProgressCallback progress_callback
+                     cancellable, // gpointer progress_callback_data
+                     NULL, // GAsyncReadyCallback callback
+                     NULL); // gpointer user_data
```
What are these "//" comments for?

parazyd replied:
Removed. Leftover when I needed them for reference.

spinal84 commented on debian/rules:
```diff
@@ -1,114 +1,50 @@
+export DEB_LDFLAGS_MAINT_APPEND  = $(LDFLAGS_MAINT)
 
-	touch build-stamp
+%:
+	dh $@
```
Let's move this part to the top of file, like it's done in dh manpage.
I think all commented lines in the head of file may be removed.

parazyd replied:
I would rather keep it as it is, having the if logic and exports before the actual target. It makes more sense whe looking at it procedurally.

spinal84 commented on src/util.cc:
```diff
@@ -3351,79 +3359,75 @@ call_copy_cont (GnomeVFSResult result)
-
-      gnome_vfs_async_cancel (handle);
-    }
+  if (cur == tot)  // Implies the copy is done?
```
Is this comment a subject to remove?

parazyd replied:
Removed.

spinal84 commented:
You may squash the commit "Fix call_copy_cont() poor logic" -> "Port to GIO"

parazyd committed:
Update debian/changelog.

MerlijnWajer commented:
Thank you all for the extensive reviews, and Parazyd for doing the work.

spinal84 commented:
@MerlijnWajer, to early... it was a WIP branch.

MerlijnWajer commented:
Parazyd told me it was ready and asked me to merge it, if I understand it correctly.

spinal84 commented:
Still a lot of issues remain.
Will have to do a new PR.
